examples/main/main.cpp

```diff
@@ -428,7 +430,12 @@ int main(int argc, char ** argv) {
                     last_n_tokens.data() + last_n_tokens.size() - last_n_repeat,
                     last_n_repeat, alpha_frequency, alpha_presence);
                 if (!penalize_nl) {
-                    logits[llama_token_nl()] = nl_logit;
+                  for (size_t idx = 0; idx < candidates_p.size; idx++) {
+                    if (candidates_p.data[idx].id == llama_token_nl()) {
+                      candidates_p.data[idx].logit = nl_logit;
+                      break;
+                    }
+                  }
```

Comment:
Hmm, the order of `candidates_p` does not seem to be changed.
Why do you think it has?

Reply:
It's not that the order of `candiates_p` has changed, although IMO it's not great to rely on the fact that `llama_sample_repetition_penalty` and `llama_sample_frequency_and_presence_penalties` preserve the order (without at least documenting that). The main issue is that on line 418 we've copied the logits into local candidates vector (of llama_token_data_array, not raw floats). So modifying the original logit array from the context does nothing, right?

Reply:
@tom7 Apologise for the delay and not paying more attention to this - I think you are right and we've had this bug for quite some time now.